### DISCUSSION TITLE
TOO_MANY_DATA_ELEMENTS when repeating a segment

### DISCUSSION THREAD

**Discussion by xander-proact** [xlate/staedi#494] [Q&A] [answered]
I have created several schema files for EDIFACT and when repeating a composite type, I am faced with an error of "TOO_MANY_DATA_ELEMENTS"

I am repeating a composite C212 within a PIA segment; the first occurrence works correctly but subsequent occurrences result in the stated error

Here is the schema entry for the segment: 

```
	<segmentType name="PIA">
		<sequence>
			<element type="E4347" minOccurs="1" maxOccurs="1"/>
			<composite type="C212" minOccurs="1" maxOccurs="5"/>
		</sequence>
	</segmentType>

```
It does not matter whether I include minOccurs or maxOccurs, it only accepts a single instance of the composite C212. This is just an example, I have had issues with similar composites in other schema files

the composite C212 is as follows

```
	<compositeType name="C212">
		<sequence>
			<element type="E7140" minOccurs="1" maxOccurs="1"/>
			<element type="E7143" minOccurs="1" maxOccurs="1"/>
			<element type="E1131"/>
			<element type="E3055"/>
		</sequence>
	</compositeType>
```

The elements E1131 and E3055 are not currently used in this composite but have been added to the schema to ensure completeness. 

E7140 is a basic string:

`	<elementType name="E7140" base="string" maxLength="35"/>`

and E7143 is a string with enumeration. I have anonymised the values here. these appear to evaluate correctly, and each is used a maximum of one time in the generated file:

```
	<elementType name="E7143" base="string" maxLength="3">
		<enumeration>
			<value>X01</value>
			<value>X02</value>
			<value>X03</value>
			<value>X04</value>
			<value>X05</value>
		</enumeration>
	</elementType>
```

Have I overlooked some aspect of the schema here in order to allow repeated occurrences within a segment?

**Comment by MikeEdgar** [comment]
Hi @xander-proact , it seems like what you have in your schema should work. Do you have a sample `PIA` segment you can share as well?

**Reply by xander-proact** [reply]
Hi @MikeEdgar , Thanks for taking a look. The PIA looks like this: 

`PIA+1+ABC:X01+XYZ:X02+DEF:X03+001:X04'`

**Reply by MikeEdgar** [reply]
How are you writing the segment? I would expect to see the repetition separator `*` between occurrences of `C212` rather than the element delimiter `+`, like this:
```
PIA+1+ABC:X01*XYZ:X02*DEF:X03*001:X04'
```

That's assuming the default EDIFACT repetition character `*`.

**Reply by MikeEdgar** [reply, marked as answer]
The error `TOO_MANY_DATA_ELEMENTS` is because the validator "sees" element PIA03, but what you want are multiple occurrences of PIA02. The repeater `*` should do that.

**Reply by xander-proact** [reply]
I can't find * as the default repetition separator in the UNECE specification, the chapter on repetition (8.2 - https://unece.org/fileadmin/DAM/trade/edifact/untdid/d422_s.htm#repdata ) does not prescribe the use of a specific character for element repetition, instead gives examples of using the + or holding it within a composite as non-mandatory to use : for repetition - where have you seen this specified as the default?

**Reply by MikeEdgar** [reply]
It looks like that link is dealing with version 3, which I do not believe supports repeating elements. For version 4, see document [ISO 9735-1:2002 Part 1: Syntax rules common to all parts](https://service.gefeg.com/jwg1/Files/V41-9735-1.pdf) from here [1]. See the table on page 22 (labelled page 14 in the PDF) for the default delimiters when the `UNA` header segment is not specified.

[1] https://service.gefeg.com/jwg1/Current/dx1.htm

**Comment by xander-proact** [comment]
While upgrading the specification to Syntax Version 4 does indeed fix this issue, due to real-world limitations with third party service providers it's not as straightforward as just upgrading the specification.

I have managed to validate Syntax Version 3 successfully by simply repeating the offending composite element in the schema XML file the required number of times, so each is technically a separate element within the segment (treating the repetitions as PIA02, PIA03, PIA04 etc): 


```
	<segmentType name="PIA">
		<sequence>
			<element type="E4347" minOccurs="1" maxOccurs="1"/>
			<composite type="C212"/>
			<composite type="C212"/>
			<composite type="C212"/>
			<composite type="C212"/>
			<composite type="C212"/>
		</sequence>
	</segmentType>
```

Marked the response to change the syntax version as solution as this is technically the correct procedure; closing issue
